### PR TITLE
[tpcds](nereids) add rule to eliminate empty relation

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes
1. eliminate emptyrelation, 
2. const fold after filter pushdown

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.42 seconds
 stream load tsv:          542 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17165981792 Bytes

**Comment by englefly** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.3 seconds
 stream load tsv:          532 seconds loaded 74807831229 Bytes, about 134 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17166327359 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        return ImmutableList.of(
+            // join->empty
+            logicalJoin(any(), any())
+                .when(this::hasEmptyRelationChild)
+                .then(join -> {
+                    if (canReplaceJoinByEmptyRelation(join)
+                            || (join.getJoinType() == JoinType.LEFT_OUTER_JOIN
+                                && join.left() instanceof EmptyRelation)
+                            || (join.getJoinType() == JoinType.RIGHT_OUTER_JOIN
+                                && join.right() instanceof EmptyRelation)) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java:55]: put these condition into function `canReplaceJoinByEmptyRelation`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                if (union.child(0) instanceof EmptyRelation) {
+                    child = union.child(1);
+                } else if (union.child(1) instanceof EmptyRelation) {
+                    child = union.child(0);
+                } else {
+                    child = null;
+                }
+                if (child == null) {
+                    return null;
+                }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/EliminateEmptyRelation.java:85]: it is better to process all match EmptyRelation here too

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.82 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17164985798 Bytes

**Review by Kikyou1997 — APPROVED** [review]

**Review by starocean999 — APPROVED** [review]

**Pull request merged**